xtask/src/flash_image.rs

```diff
@@ -0,0 +1,504 @@
+// Licensed under the Apache-2.0 license
+
+use crate::DynError;
+use crc32fast::Hasher;
+use std::fs::File;
+use std::io::{self, Read, Write};
+use std::io::{Error, ErrorKind};
+
+const FLASH_IMAGE_MAGIC_NUMBER: u32 = 0x464C5348;
+const HEADER_VERSION: u16 = 0x0001;
+const CALIPTRA_FMC_RT_IDENTIFIER: u32 = 0x00000001;
+const SOC_MANIFEST_IDENTIFIER: u32 = 0x00000002;
+const MCU_RT_IDENTIFIER: u32 = 0x00000002;
+const SOC_IMAGES_BASE_IDENTIFIER: u32 = 0x00001000;
+
+pub struct FlashImage {
+    header: FlashImageHeader,
+    checksum: FlashImageChecksum,
+    payload: FlashImagePayload,
+}
+
+pub struct FlashImageHeader {
+    magic_number: u32,
+    header_version: u16,
+    image_count: u16, // number of images
+}
+
+pub struct FlashImageChecksum {
+    header: u32,  // checksum of the header
+    payload: u32, // checksum of the payload
+}
+
+pub struct FlashImagePayload {
+    image_info: Vec<FlashImageInfo>,
+    images: Vec<FirmwareImage>,
+}
+
+// Per image header
+pub struct FlashImageInfo {
+    identifier: u32,
+    image_offset: u32, // Location of the image in the flash as an offset from the header
+    size: u32,         // Size of the image
+}
+
+#[derive(Clone)]
+pub struct FirmwareImage {
+    identifier: u32,
+    data: Vec<u8>,
+}
+
+impl FirmwareImage {
+    pub fn new(identifier: u32, filename: &str) -> io::Result<Self> {
+        let mut file = File::open(filename)?;
```

Comment:
It would be useful to have an error message printed here so that we know which file failed to open when running this.

Reply:
ack